File: SDO/SDO.APP/RPT/Word/ProjectOther.cs
using Aspose.Words;
using Autofac;
using SDO.Base.RPT.Enums;
using SDO.Models;
using SDO.ReportBuilder.Services;
using SDO.Services;
using SDO.Utils;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SDO.APP.RPT.Word
{
    /// <summary>
    /// 預覽列印 三、執行情形 （五）其它資訊
    /// </summary>
    public class ProjectOther : WContentBuilder
    {
        private readonly IProjectExecuteService projectExecuteService;
        // 其它資訊資料
        ProjectFillOtherModel projectOtherData;
        List<bool> checkList;
        int count = 0;

        public ProjectOther(IComponentContext coms) : base(coms)
        {
            this.projectExecuteService = coms.Resolve<IProjectExecuteService>();
        }

        protected override async Task GetData()
        {
            projectOtherData = await projectExecuteService.GetProjectFillOther(Parameter.PROJECT_NO);
            checkList = (List<bool>)Parameter.ObjectModel;
        }

        protected override void Content()
        {
            Bid();
            Activity();
            Review();
            Tender();
        }

        /// <summary>
        /// 招標情形
        /// </summary>
        private void Bid()
        {
            if (checkList[0])
            {
                SetTitle($"{++count}.招標情形", fontSize: 14);
                foreach (ProjectBidModel item in projectOtherData.ProjectBid)
                {
                    if (projectOtherData.ProjectBidDetail.Where(x => x.BID_KIND == item.BID_KIND).Any() ||
                        item.AWARD_BID_DATE.HasValue ||
                        !string.IsNullOrEmpty(item.BID_TENDER))
                    {
                        Table = Builder.StartTable();
                        SetThColumn(item.BID_NAME, backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center, hMergeCount: 3);
                        Builder.EndRow();

                        SetThColumn("決標日期", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                        SetTdColumn(item.AWARD_BID_DATE.ToTwDateString(), hMergeCnt: 2);
                        Builder.EndRow();

                        SetThColumn("決標廠商", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                        SetTdColumn(item.BID_TENDER, hMergeCnt: 2);
                        Builder.EndRow();

                        SetThColumn("流標次數", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                        SetThColumn("流標日期", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                        SetThColumn("流標原因", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                        Builder.EndRow();
                        BidDetail(projectOtherData.ProjectBidDetail.Where(x => x.BID_KIND == item.BID_KIND && x.DETAIL_TYPE == 0).ToList());

                        SetThColumn("廢標次數", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                        SetThColumn("廢標日期", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                        SetThColumn("廢標原因", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                        Builder.EndRow();
                        BidDetail(projectOtherData.ProjectBidDetail.Where(x => x.BID_KIND == item.BID_KIND && x.DETAIL_TYPE == 1).ToList());
                        Builder.EndTable();
                        InitTable(Table, new List<double> { 15, 20, 65 });
                        Builder.Writeln(string.Empty);
                    }
                }
            }
        }

        /// <summary>
        /// 流標/廢標歷程
        /// </summary>
        /// <param name="data"></param>
        private void BidDetail(List<ProjectBidDetailModel> data)
        {
            if (!data.Any())
            {
                SetTdColumn("無資料", alignment: AlignmentEnum.Center, hMergeCnt: 3);
                Builder.EndRow();
            }
            int count = 1;
            foreach (ProjectBidDetailModel item in data)
            {
                SetTdColumn(count, alignment: AlignmentEnum.Center);
                SetTdColumn(item.DETAIL_DATE.ToTwDateString(), alignment: AlignmentEnum.Center);
                SetTdColumn(item.DETAIL_REASON);
                Builder.EndRow();
                count++;
            }
        }

        /// <summary>
        /// 相關活動
        /// </summary>
        private void Activity()
        {
            if (checkList[1])
            {
                SetTitle($"{++count}.相關活動", fontSize: 14);
                Table = Builder.StartTable();
                foreach (ProjectActivityModel item in projectOtherData.ProjectActivity)
                {
                    SetThColumn(item.ACTIVITY_KIND_NAME, backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Left);
                    if (item.ACTIVITY_KIND == "04" && item.editType == 0)
                    {
                        SetTdColumn("無資料", hMergeCnt: 3);
                        Builder.EndRow();
                    }
                    else
                    {
                        string activity = item.ACTIVITY_KIND == "04"
                            ? item.ACTIVITY_NAME
                            : !item.IS_ACTIVITY.HasValue ? "" : item.IS_ACTIVITY.Value ? "有" : "無";
                        SetTdColumn(activity);
                        SetThColumn("活動日期", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Left);
                        SetTdColumn(item.ACTIVITY_DATE.ToTwDateString());
                        Builder.EndRow();
                    }
                }
                Builder.EndTable();
                InitTable(Table, new List<double> { 30, 35, 15, 20 });
                Builder.Writeln(string.Empty);
            }
        }

        /// <summary>
        /// 相關審查
        /// </summary>
        private void Review()
        {
            if (checkList[2])
            {
                SetTitle($"{++count}.相關審查", fontSize: 14);
                Table = Builder.StartTable();
                foreach (SetParamModel item in projectOtherData.SetParam)
                {
                    foreach (ProjectReviewModel review in projectOtherData.ProjectReview.Where(x => x.REVIEW_KIND == item.SET_TYPE).ToList())
                    {
                        SetThColumn(item.SET_VALUE, backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Left);
                        string value = review.REVIEW_KIND != "04"
                                   ? !review.IS_REVIEW.HasValue ? "" : review.IS_REVIEW.Value ? "有" : "無"
                                   : review.OTH_RVWNAME;
                        SetTdColumn(value);
                        SetThColumn("送件日期", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Left);
                        SetTdColumn(review.SEND_DATE.ToTwDateString());
                        SetThColumn("核定日期", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Left);
                        SetTdColumn(review.REVIEW_DATE.ToTwDateString());
                        Builder.EndRow();
                    }
                    if (item.SET_TYPE == "04" && !projectOtherData.ProjectReview.Where(x => x.REVIEW_KIND == "04").ToList().Any())
                    {
                        SetThColumn(item.SET_VALUE, backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Left);
                        SetTdColumn("無資料", hMergeCnt: 5);
                        Builder.EndRow();
                    }
                }
                Builder.EndTable();
                InitTable(Table, new List<double> { 30, 14, 14, 14, 14, 14 });
                Builder.Writeln(string.Empty);
            }
        }

        /// <summary>
        /// 廠商資訊
        /// </summary>
        private void Tender()
        {
            if (checkList[3])
            {
                SetTitle($"{++count}.廠商資訊", fontSize: 14);
                Table = Builder.StartTable();
                SetThColumn("廠商類別", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                SetThColumn("廠商名稱", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                SetThColumn("統編", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                SetThColumn("地址", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                SetThColumn("聯絡人", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                SetThColumn("聯絡人電話", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                SetThColumn("聯絡人Email", backGroundColor: Color.FromArgb(222, 234, 246), alignment: ParagraphAlignment.Center);
                Builder.EndRow();

                foreach (ProjectTenderModel item in projectOtherData.ProjectTender)
                {
                    SetTdColumn(item.TENDER_KIND_NAME);
                    SetTdColumn(item.TENDER_NAME);
                    SetTdColumn(item.REG_NO);
                    SetTdColumn(item.TENDER_ADDR);
                    SetTdColumn(item.CONTACT_NAME);
                    SetTdColumn(item.CONTACT_PHONE);
                    SetTdColumn(item.CONTACT_EMAIL);
                    Builder.EndRow();
                }
                Builder.EndTable();
                InitTable(Table, new List<double> { 12, 16, 12, 22, 9, 14, 15 });
            }
        }
    }
}

File: Front/PWS/src/Page/PWS/ProjectPrint/ProjectOther.js
import React from 'react';
import { IsNullOrEmpty, FormatDate } from '../../../Basic/SDOExtension';
import TextAreaWrapInput from '../../../Components/Input/TextAreaWrapInput';

// 三、執行情形（五）其它資訊
const ProjectOther = (props) => {
    const { projectOther, title, checkList } = props;
    let count = 0;

    // 日期格式化
    const formatDate = (date) => {
        if (IsNullOrEmpty(date)) {
            return ""
        }
        else {
            return FormatDate(date, "tYY/MM/DD")
        }
    }

    // 招標情形
    const bid = () => {
        let result = projectOther.ProjectBid.map(x => {
            let detailType0Data = projectOther.ProjectBidDetail.filter(y => y.BID_KIND == x.BID_KIND && y.DETAIL_TYPE == 0);
            let detailType1Data = projectOther.ProjectBidDetail.filter(y => y.BID_KIND == x.BID_KIND && y.DETAIL_TYPE == 1);
            if (!IsNullOrEmpty(x.AWARD_BID_DATE) || !IsNullOrEmpty(x.BID_TENDER) ||
                detailType0Data.length != 0 || detailType1Data.length != 0) {
                return (
                    <form>
                        <table style={{ wordBreak: "break-all" }}>
                            <tr>
                                <th colSpan={3} style={{ textAlign: "center" }}>{x.BID_NAME}</th>
                            </tr>
                            <tr>
                                <th style={{ width: "15%" }}>決標日期</th>
                                <td colSpan={2}>{formatDate(x.AWARD_BID_DATE)}</td>
                            </tr>
                            <tr>
                                <th>決標廠商</th>
                                <td colSpan={2}>{x.BID_TENDER}</td>
                            </tr>
                            <tr>
                                <th style={{ textAlign: "center" }}>流標次數</th>
                                <th style={{ textAlign: "center", width: "20%" }}>流標日期</th>
                                <th style={{ textAlign: "center" }}>流標原因</th>
                            </tr>
                            {bidDetail(detailType0Data)}
                            <tr>
                                <th style={{ textAlign: "center" }}>廢標次數</th>
                                <th style={{ textAlign: "center" }}>廢標日期</th>
                                <th style={{ textAlign: "center" }}>廢標原因</th>
                            </tr>
                            {bidDetail(detailType1Data)}
                        </table>
                    </form>
                )
            }
        });
        return result;
    }

    // 流標/廢標歷程
    const bidDetail = (data) => {
        if (data.length == 0) {
            return (
                <tr>
                    <td colSpan={3} style={{ textAlign: "center" }}>無資料</td>
                </tr>
            )
        }
        return data.map((x, index) => {
            return (
                <tr>
                    <td style={{ textAlign: "center" }}>{index + 1}</td>
                    <td style={{ textAlign: "center" }}>{formatDate(x.DETAIL_DATE)}</td>
                    <td><TextAreaWrapInput value={x.DETAIL_REASON} /></td>
                </tr>
            )
        })
    }

    // 相關活動
    const activity = () => {
        return <form>
            <table style={{ wordBreak: "break-all" }}>
                {projectOther.ProjectActivity.map(x => {
                    return (
                        <tr>
                            <th>{x.ACTIVITY_KIND_NAME}</th>
                            {x.ACTIVITY_KIND == 4 && x.editType == 0
                                ? <td colSpan={3}>無資料</td>
                                :
                                <>
                                    <td>{x.ACTIVITY_KIND != 4
                                        ? IsNullOrEmpty(x.IS_ACTIVITY) ? "" : x.IS_ACTIVITY ? "有" : "無"
                                        : x.ACTIVITY_NAME}
                                    </td>
                                    <th>活動日期</th>
                                    <td>{formatDate(x.ACTIVITY_DATE)}</td>
                                </>
                            }
                        </tr>
                    )
                })}
            </table>
        </form>;
    }

    // 相關審查
    const review = () => {
        const content = () => {
            return projectOther.SetParam.map(x => {
                return (
                    <>
                        {projectOther.ProjectReview.filter(y => y.REVIEW_KIND == x.SET_TYPE).map((y) => {
                            return (
                                <tr>
                                    <th style={{ width: "18%" }}>{x.SET_VALUE}</th>
                                    <td style={{ width: "18%" }}>
                                        {y.REVIEW_KIND != "04"
                                            ? IsNullOrEmpty(y.IS_REVIEW) ? "" : y.IS_REVIEW ? "有" : "無"
                                            : y.OTH_RVWNAME}
                                    </td>
                                    <th>送件日期</th>
                                    <td>{formatDate(y.SEND_DATE)}</td>
                                    <th>核定日期</th>
                                    <td>{formatDate(y.REVIEW_DATE)}</td>
                                </tr>
                            )
                        })}
                        {x.SET_TYPE == "04" && projectOther.ProjectReview.filter(y => y.REVIEW_KIND == "04").length == 0 &&
                            <tr>
                                <th>{x.SET_VALUE}</th>
                                <td colSpan={5}>無資料</td>
                            </tr>
                        }
                    </>
                )
            })
        }

        return <form><table style={{ wordBreak: "break-all" }}>{content()}</table></form>;
    }

    // 廠商資訊
    const tender = () => {
        return (
            <form>
                <table style={{ wordBreak: "break-all" }}>
                    <tr>
                        <th style={{ textAlign: "center", width: "12%" }}>廠商類別</th>
                        <th style={{ textAlign: "center" }}>廠商名稱</th>
                        <th style={{ textAlign: "center", width: "7%" }}>統編</th>
                        <th style={{ textAlign: "center" }}>地址</th>
                        <th style={{ textAlign: "center" }}>聯絡人</th>
                        <th style={{ textAlign: "center" }}>聯絡人電話</th>
                        <th style={{ textAlign: "center" }}>聯絡人Email</th>
                    </tr>
                    {projectOther.ProjectTender.length == 0 &&
                        <tr>
                            <td colSpan={7} style={{ textAlign: "center" }}>無資料</td>
                        </tr>
                    }
                    {projectOther.ProjectTender.map(x =>
                        <tr>
                            <td>{x.TENDER_KIND_NAME}</td>
                            <td>{x.TENDER_NAME}</td>
                            <td>{x.REG_NO}</td>
                            <td>{x.TENDER_ADDR}</td>
                            <td>{x.CONTACT_NAME}</td>
                            <td>{x.CONTACT_PHONE}</td>
                            <td>{x.CONTACT_EMAIL}</td>
                        </tr>
                    )}
                </table>
            </form>
        )
    }

    return (
        <>
            <p>{title}</p>
            {checkList[0] &&
                <>
                    <p>{++count}.招標情形</p>
                    {bid()}
                </>
            }

            {checkList[1] &&
                <>
                    <p>{++count}.相關活動</p>
                    {activity()}
                </>
            }

            {checkList[2] &&
                <>
                    <p>{++count}.相關審查</p>
                    {review()}
                </>
            }

            {checkList[3] &&
                <>
                    <p>{++count}.廠商資訊</p>
                    {tender()}
                </>
            }
        </>
    );
}

export default ProjectOther;